### 百度API：通用物体和场景识别

[官方说明文档](https://ai.baidu.com/ai-doc/IMAGERECOGNITION/Xk3bcxe21)

[鉴权认证机制(获取Access Token)](https://ai.baidu.com/ai-doc/REFERENCE/Ck3dwjhhu)

**HTTP 方法：** POST

**请求URL：** https://aip.baidubce.com/rest/2.0/image-classify/v2/advanced_general

In [1]:
# A-1 获取access_token等参数信息
# encoding:utf-8
# client_id 为官网获取的AK， client_secret 为官网获取的SK
import requests 
import pandas as pd
import json

host = 'https://aip.baidubce.com/oauth/2.0/token'
params = {
    "grant_type":"client_credentials", # grant_type=client_credentials&client_id
    "client_id":"whLpOu1PE8jIv6BSGMwqbBa3", # 【官网获取的AK】
    "client_secret":"MAr7YkeVlyV2Bo7B2MGc3T5rdpaSII16" # 【官网获取的SK】
}
response = requests.get(host,params=params)
display(pd.json_normalize(response.json()).T) # T改变为上下排列

access_token=response.json()["access_token"]
access_token

0
refresh_token   25.f673a23f55b8ac2b9f862124fea25095.315360000....
expires_in                                                2592000
session_key     9mzdWrrydc0Z0Tz3SqztkrQaZ49x8bsXdQKRFCphBzPEzX...
access_token    24.af8c5c0f2b2fbdc2e5b5d885f522ce13.2592000.16...
scope           public vis-classify_dishes vis-classify_car br...
session_secret                   4d02c8e1852899df1131513c4ab78e4d

'24.af8c5c0f2b2fbdc2e5b5d885f522ce13.2592000.1612944227.282335-23525275'

In [4]:
# encoding:utf-8

# AppID = '23525275'
# API Key = 'whLpOu1PE8jIv6BSGMwqbBa3'
# Secret Key = 'MAr7YkeVlyV2Bo7B2MGc3T5rdpaSII16'


import requests
import base64

'''
通用物体和场景识别
'''

request_url = "https://aip.baidubce.com/rest/2.0/image-classify/v2/advanced_general"
# 二进制方式打开图片文件
f = open('ptest05.jpg', 'rb')
img = base64.b64encode(f.read())

params = {"image":img}
access_token = '24.af8c5c0f2b2fbdc2e5b5d885f522ce13.2592000.1612944227.282335-23525275'
request_url = request_url + "?access_token=" + access_token
headers = {'content-type': 'application/x-www-form-urlencoded'}
response = requests.post(request_url, data=params, headers=headers)
# if response:
#     print (response.json())
    
result = response.json()
result

{'log_id': 5712521386408023851,
 'result_num': 5,
 'result': [{'score': 0.358338, 'root': '建筑-其他', 'keyword': '居民楼'},
  {'score': 0.253114, 'root': '自然风景-天空', 'keyword': '天空'},
  {'score': 0.17113, 'root': '建筑-现代建筑', 'keyword': '剧院/博物馆/礼堂'},
  {'score': 0.088515, 'root': '自然风景-其他', 'keyword': '风景'},
  {'score': 0.006694, 'root': '植物-树', 'keyword': '树'}]}

In [1]:
# encoding:utf-8

# AppID = '23525275'
# API Key = 'whLpOu1PE8jIv6BSGMwqbBa3'
# Secret Key = 'MAr7YkeVlyV2Bo7B2MGc3T5rdpaSII16'


import requests
import base64

'''
通用物体和场景识别
'''

request_url = "https://aip.baidubce.com/rest/2.0/image-classify/v2/advanced_general"
# 二进制方式打开图片文件
f = open('stest02.jpg', 'rb')
img = base64.b64encode(f.read())

params = {"image":img}
access_token = '24.af8c5c0f2b2fbdc2e5b5d885f522ce13.2592000.1612944227.282335-23525275'
request_url = request_url + "?access_token=" + access_token
headers = {'content-type': 'application/x-www-form-urlencoded'}
response = requests.post(request_url, data=params, headers=headers)
# if response:
#     print (response.json())
    
result = response.json()
result

{'log_id': 6639343850218473740,
 'result_num': 5,
 'result': [{'score': 0.31101, 'root': '商品-电脑办公', 'keyword': '笔记本电脑'},
  {'score': 0.230072, 'root': '商品-数码产品', 'keyword': '台式电脑'},
  {'score': 0.155608, 'root': '商品-家用电器', 'keyword': '音箱'},
  {'score': 0.076702, 'root': '商品-电脑办公', 'keyword': '键盘'},
  {'score': 0.001803, 'root': '商品-电脑办公', 'keyword': '电脑'}]}

#### 返回参数说明

|参数|返回值|说明|
| --- | --- | --- |
|log_id|6639343850218473740|用于问题定位|
|result_num|5|返回结果数目，及result数组中的元素个数|
|score|(0.31101)|置信度，0-1|
|root|商品|识别结果的上层标签|
|keyword|(笔记本电脑)|图片中的物体或场景名称|